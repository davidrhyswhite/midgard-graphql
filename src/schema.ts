import { join } from 'path';
import { loadSchemaSync } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { addResolversToSchema } from '@graphql-tools/schema';
import resolvers from './resolvers';

const schema = loadSchemaSync(join(__dirname, './schema/schema.graphql'), {
  loaders: [new GraphQLFileLoader()],
});

const schemaWithResolvers = addResolversToSchema({
  schema,
  resolvers,
});

export default schemaWithResolvers;
