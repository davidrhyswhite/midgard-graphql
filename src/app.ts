import { Config } from 'apollo-server-core';
import { ApolloServer as Server } from 'apollo-server';
import { ApolloServer as Lambda } from 'apollo-server-lambda';
import { loadConfigSync } from 'graphql-config';
import schema from './schema';
import MidgardAPI from './datasources/midgard';
import { DataSources } from './types/context';

const config = loadConfigSync({ rootDir: './' });

interface ApolloConstructable<T> {
  new (args: Config): T;
}
type ServerType = Server | Lambda;

export function createServer(App: ApolloConstructable<ServerType>): ServerType {
  return new App({
    schema,
    playground: {
      config,
    },
    introspection: true,
    dataSources: (): DataSources => ({
      midgardAPI: new MidgardAPI(),
    }),
  });
}
