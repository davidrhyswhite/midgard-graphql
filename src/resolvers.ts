import { Context } from './types/context';
import { Stats } from './types/midgard/stats';
import { IResolvers } from 'apollo-server';
import { AssetPool } from './types/midgard/asset-pools';

const resolvers: IResolvers = {
  Query: {
    stats: (_: void, __: void, { dataSources }: Context): Promise<Stats> => dataSources.midgardAPI.getStats(),
    pools: (_: void, __: void, { dataSources }: Context): Promise<Array<AssetPool>> =>
      dataSources.midgardAPI.getAssetPools(),
  },
};

export default resolvers;
