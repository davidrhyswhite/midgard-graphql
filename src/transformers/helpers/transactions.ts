import { assetFromString } from '@thorchain/asgardex-util';
import {
  TransactionDetail,
  Transaction,
  Sum,
  MidgardTransactionDetail,
  MidgardTx,
  MidgardSum,
} from 'midgard/transactions';

export function toSum(sum: MidgardSum): Sum {
  return {
    ...sum,
    asset: assetFromString(sum.asset),
  };
}

export function toGas(sum: MidgardSum): Sum | void {
  if (typeof sum === 'undefined') {
    return null;
  }
  return toSum(sum);
}

export function toTransaction(data: MidgardTx): Transaction {
  return {
    ...data,
    coins: data.coins.map(toSum),
  };
}

export function toTransactionsDetail(data: MidgardTransactionDetail): TransactionDetail {
  return {
    ...data,
    pool: assetFromString(data.pool),
    in: toTransaction(data.in),
    out: data.out.map(toTransaction),
    gas: toGas(data.gas),
  };
}
