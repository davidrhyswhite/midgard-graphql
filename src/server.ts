import app from './app';
import config from 'config';

const port = config.get('app.port') as number;

app.listen({ port }, (): void => {
  console.log(`\nðŸš€ GraphQL is now running on http://localhost:${port}/graphql`);
});
