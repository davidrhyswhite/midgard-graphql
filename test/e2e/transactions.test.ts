import { createTestClient } from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server';
import gql from 'graphql-tag';
import nock from 'nock';
import { createTransactions } from '../helpers/mock-creators';
import { createServer } from '../../src/app';

const GET_TRANSACTIONS = gql`
  fragment transaction on Transaction {
    txID
    address
  }
  query getTransactions {
    transactions(limit: 20, offset: 0) {
      totalCount
      nodes {
        type
        in {
          ...transaction
        }
        out {
          ...transaction
        }
      }
      pageInfo {
        nextCursor
        endCursor
        hasNextPage
      }
    }
  }
`;

test('returns an TransactionConnection type from application', async () => {
  const { query } = createTestClient(createServer(ApolloServer));

  nock('http://midgard.hostname.local')
    .get('/transactions/path')
    .query({ limit: 20, offset: 0 })
    .reply(200, createTransactions());

  // run query against the server and snapshot the output
  const res = await query({ query: GET_TRANSACTIONS });
  expect(res).toMatchSnapshot();
});
