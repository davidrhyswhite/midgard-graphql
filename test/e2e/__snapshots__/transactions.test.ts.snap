// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns an TransactionConnection type from application 1`] = `
Object {
  "data": Object {
    "transactions": Object {
      "nodes": Array [
        Object {
          "in": Object {
            "address": "address",
            "txID": "txID",
          },
          "out": Array [
            Object {
              "address": "address",
              "txID": "txID",
            },
          ],
          "type": "stake",
        },
      ],
      "pageInfo": Object {
        "endCursor": -19,
        "hasNextPage": false,
        "nextCursor": null,
      },
      "totalCount": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
